### DISCUSSION TITLE
Input Binding with Vue 3

### DISCUSSION THREAD

**Discussion by hasan-ozbey** [simeydotme/svelte-range-slider-pips#156] [Q&A] [answered]
I'm using this repo with Vue 3 and TypeScript to create a range slider with two corresponding input fields (for minimum and maximum values). When I move the slider handles, the input fields update correctly. However, when I manually change the input values, the slider does not reflect this change.

It seems like the slider does not detect external changes to the min and max values once it’s initialized. I couldn’t find any method in documentation to programmatically update the slider position based on input values. Is there a built-in way to synchronize the slider with input changes?

```typescript
<script setup lang="ts">
import { ref, reactive, onMounted } from 'vue'

const rangeDefinitions = {
  power: { min: 0, max: 81, step: 10 }
}

const rangeStates = reactive({
  power: { min: rangeDefinitions.power.min, max: rangeDefinitions.power.max }
})

const $powerRangeSliderNode = ref(null)

onMounted(async () => {
  const { default: RangeSlider } = await import('svelte-range-slider-pips')

  const powerRangeSlider = new RangeSlider({
    target: $powerRangeSliderNode.value,
    props: {
      values: [rangeStates.power.min, rangeStates.power.max],
      pips: true,
      range: true,
      float: true,
      step: rangeDefinitions.power.step,
      min: rangeDefinitions.power.min,
      max: rangeDefinitions.power.max,
      suffix: 'k'
    }
  })

  powerRangeSlider.$on('change', (e) => {
    rangeStates.power.min = e.detail.values[0]
    rangeStates.power.max = e.detail.values[1]
  })
})
</script>

<template>
  <div>
    <h4>Engine Power (kW)</h4>
    <div class="flex items-center gap-4 mt-4">
      <!-- Minimum Value Input -->
      <UFormGroup label="Minimum" name="engine-power-minimum">
        <UInput
          v-model="rangeStates.power.min"
          type="number"
          :min="rangeDefinitions.power.min"
          :max="rangeStates.power.max"
          :step="rangeDefinitions.power.step"
          :placeholder="rangeStates.power.min.toString()"
        />
      </UFormGroup>

      <!-- Maximum Value Input -->
      <UFormGroup label="Maximum" name="engine-power-maximum">
        <UInput
          v-model="rangeStates.power.max"
          type="number"
          :min="rangeStates.power.min"
          :max="rangeDefinitions.power.max"
          :step="rangeDefinitions.power.step"
          :placeholder="rangeStates.power.max.toString()"
        />
      </UFormGroup>
    </div>

    <!-- Slider Element -->
    <div id="power-slider" ref="$powerRangeSliderNode" class="mt-8 px-2" />
  </div>
</template>
```

Moving the slider updates `rangeStates.power.min` and `rangeStates.power.max`, which reflects in the `UInput` fields. However, manually updating the `UInput` values does not move the slider handles accordingly.

Any guidance on how to sync the slider with manual input changes would be appreciated!

**Comment by hasan-ozbey** [comment]
Forgot to mention that if I import this package synchronously (`import RangeSlider from 'svelte-range-slider-pips'`) I get this error on Nuxt while using `new RangeSlider()`

"Class extends value undefined is not a constructor or null"

**Reply by simeydotme** [reply]
This i would need a Nuxt/Vue expert to help understand, I really have very minimal experience with Nuxt, and not much working experience of Vue3 🙇

**Comment by simeydotme** [comment]
Here's [a working version](https://play.vuejs.org/#eNq1Vu9v2zYQ/VcIr4MdwJLjpO0w1QmadQXWYW2Hdt/ifJCls8SEIgWSsuMG/t/3+MO2kiVd+6EOYkvHd+/u3h0p3Q0u2jZddTTIBjNTaN5aZsh27flcziVjvGmVtuyOaVqOmZLvVSctlWzLllo1bAjP4ase8FMuK/oseEk6IsyKhKVEu4XE+JWk5a0ZvmKTCStqKm6YrSkSGNbkbQhdKGksW+WiI8POXAKjy9PTMXv58urI+8Ylq9iCy9L9BvqDc7iPzrITIjiWyAsW0iQLYkulPQGXFQpkjavQUQiyrNnEWh5SRGaOKLkj2dOF3PdCjUZH7Oyc3Tkj845WaWJ5Lz7ydvVHxjW3tb8f2lxXZIcHT+TjFtx/xVck2e8f3/cDO1wvYUnrfjtGMQfGAnEWI6ZexvFusdWqNdkuYfcJMmfx9wEcDuhlxqy+Z/TdfmDdxqvtEQZmX1XXlrkNRcV+HpRZ1yjyoM3eJwJzCFnULlLJOuPad8AODXuWKjkaBsDwiDVka1XeVyl91oOM2Yj6zdpVHiqGoJSWZHMugsH4sY/l9KpadrKwHIME1lLQG8Ex4JiCyPqA83I6HZ+cXEWuQ15o9ujucel34i5zYWgf3n3N5WwS9rDfvTNLTSugLu4Ym5V85S9wWU/P73bsbLudTWCIS0AxXp7NB1HzgetH7/Z8NukRLTprUeoqUTIrXKVA9uoGPPY3xJpNgkM8XEDAZdthrDctwVN2zcKHzDwcluB2eXwF42uPhfHxNgHkexSmO2h1YT54yijv/wm8l9bNLZSNDUYV35nv9Fvynf7QfJFrbNRs0psD3Bq7Ee4y00rhZPczq6RNlnnDxSZj73Bu6TEzGwO/pONjdoHDhsP0B4kVWV7kMGmeC4ByaRJDmi99yoJLSmriVY3jZZq+8EZPvo7G58fHcZ8USiidsZ9Opqcvnv/ikYu8uKk0Ds4y2a0u/We3tRyT2Uhsc8OhglSSvKOlW5vgzIBSOAYyplrLG/6F/qKKL7jgduNhyGJxw1Gp52lQfu3hubSohueGyoBr1JdEmdv/ACudb0yRixB1R+ejG4RL8vK6M6704+OfPWQusTHnsp4GnT0SkSqZsYKczkB5xEKVm4ApuUGz0IeloFtPgtuCoAjwEuT3Hd1+9X5rXto6Y7/GyHuaSvNQVaDhaCqk25MwVuUtMtbUBEocIXFABuNBeCgneCan10ZJvCT4WPO4YOaD/aMCR8QTT3oHwrFgLR4Uk0lRSnCVJPhKp5LsRLbN5CnX1ycp/jDIxj6JSRsu0+YaYeJBuEXi1kCvJa8epF2opuWC9EcMCN4Q7qWfC6HWf3pb77kFH/eK8oj92tyGyv7WhC2wovlgvxY2dVh++/kD2t5bbFTZCaC/sviJjBKdyzHAfsOeQNo9nM/2ne8CRvMf8/YW02F2RblEvRoej3OpozdfKf2Q7ml6uldx+y8ILV/a) based on [the documentation website](https://simeydotme.github.io/svelte-range-slider-pips/en/getting-started/#-for-use-in-a-vuejs-project) link to "see a vue example running at play.vuejs"

If you have any improvements / suggestions for the documentation to help Vue users get started easier, please let me know and/or create some PRs on the `astro-pages` branch 🙏

**Reply by simeydotme** [reply]
I'm sure someone who's a Vue expert could extract this;

```vue
    <input type="number" :value="values[0]" @input="(e) => {
      values[0] = e.target.valueAsNumber;
      mySlider.$set({ values: values, range: true })
    }">
``` 

into some kind of directive / action that auto-binds to a slider.
If you / someone comes up with that, please let me know so we can add it to the Documentation 🙏

**Reply by simeydotme** [reply, marked as answer]
Ok, there's an even easier way to do it;

[see here, im using v-model](https://play.vuejs.org/#eNqVVm1v2zYQ/isHr4MdwJLipO0w1QnadQXWYV2Hdt/ifKCls8SEIgWSsuMG/u87vliWvaZFHSSW7p577p3M4+hN26brDkf5aG4KzVsLBm3XXi/kQgLwplXawiNoXE1ByQ+qkxZL2MFKqwbGZDl+NQB+YrLCz4KXqCPCrFFYTLRTJMZrkpa3ZvwKsgyKGot7sDVGAgMNa4PrQkljYc1EhwauXACTm8vLKbx8eXvmbaPKKlhyWbrvQH8wDu/RWHZCBMOS4iIJapQFwkppT8BlRQlC4zJ0FAItNNuYyylFZObkhTmSni7E3hdqMjmDq2t4dELwhlZpBDbwT3G7/CPjhtvav48t0xXa8cGS4nEK91vxNUr4/eOHoWOHGwQscTNsxyTGABCI8+gx9WWc7pWtVq3J9wG7TyhzHr9P4GRAvczB6iOh7/aJdBefdmc0MH1WXVsyG5KK/TxUZlNTkofa9DYRyKiQRe08ldAZ174DdmzgWarkZBwA4zNo0NaqPK5S+mwAmcIEh83aZx4ypoJiWqJlXASB8WMf0xlkFSZv2Vmr5FvBabqv4GgITlhvZrPpxcVtZAvliB0788RD2qHa8R4FfMiKRmXy+NXG9eG6Pws5z8LO+22fW2xaQfz0BjAv+do/0GM9u37c88FuN89IEFWEAl5eLUaxRyPXv8Hr9TwbEIWywDpRMi9ccQg5KBXB4zwEX/MsKPfmXLYdLcG2RbKTXbP0DtdJo0oUJApWN+e3JH3twSQc1owc/BjV7HtU+/Tm2aB69GrsVrjHXCtF56eDrpS0yYo1XGxzeE+ng56C2RqySzo+hTe00pxEf6BYo+UFI5HmTBCISZMY1Hzleye4xKRGXtW0xLP0hRd68k0UPj8/76dRKJ3DTxezyxfPf/HIJSvuK03HU5nstSv/iSaeyWwlLZPhND5SSfSGFh9sQptJydOy5aBayxv+Bf/Cii+54HbrYRTF8p5Tpp6nofRrD2fSUjacGSwDrlFfEmUe/gesNNuagongdU/nvRtyl7DyrjMu9fPznz1kIf2S1LNQZ48kT5XMoUBXZ0J5xFKV24ApuaFmUR9WAh88Cb0WSBUhvCTyY0M35d5uw0tb5/Br9NzTVJqHrAINp6ZS6XoSgIq1FLHGJlDS4sUBGU1H4epL6OZL74ySdBV7X4uoMItRfyDTYj1xnzrQYlRbS8dxlhWlJC4aZb7WqUSbybbJnjJ9fZHSDw2ysU9i0obLtLkjN/H42FHg1lC9Vrw6CbtQTcsF6o80IHRsHYXPhFCbP71scDuQjftH4CvyO/MQMvtHI63AGhejXhcusqB+9/lvavtASYvcCUJ/Q/kJjRKdizHAfqOdoLAHOB/te98FGs1/zbsHmg6zT8oF6qvh8XRmdPj2G6kfwr1ML/sq7v4DeU0v7A==) and an generic slider update function.

**Reply by hasan-ozbey** [reply]
oh seems like it would work, I'll give it a try as soon as possible. I wasn't aware that there is a built-in method to set slider values. Is this written somewhere in the documentation? 😅

**Reply by simeydotme** [reply]
> I wasn't aware that there is a built-in method to set slider values. Is this written somewhere in the documentation? 😅

Actually it's not documented except for that example, because it's a built-in part of Svelte component system actually. Thats why I really like authoring components in Svelte!

However it's only applicable to non-svelte environments and I haven't done a lot in regards to documenting that
It's a good spot to cover! thanks

(p.s. you're the first person I'm aware of that is using this outside of svelte! hope it is working nice and smooth for you!)

**Reply by hasan-ozbey** [reply]
For my case, I found it better to use a watcher instead of the `@input` event.

```typescript
const updateRangeSlider = (rangeSlider, values: number[]) => rangeSlider?.$set({ values })

watch(
  () => rangeStates.power,
  (value) => {
    updateRangeSlider(powerRangeSlider, [value.min, value.max])
  },
  { deep: true }
)
```

It is working perfectly. Thanks for this beautiful repository!

-------------

Just o be sure, this $set method does not fire $on('change') method, right?

**Reply by simeydotme** [reply]
> Just o be sure, this $set method does not fire $on('change') method, right?

from what I observed; No.
Change only fires when the slider was interacted with directly

> It is working perfectly. Thanks for this beautiful repository!
❤️
